### PR TITLE
[SYCL-MLIR]: Introduce InlineHeuristic class

### PR DESCRIPTION
**Pull request by etiotto**
This PR refactors the Inliner pass via introduction of a `InlineHeuristic` class. It also adds an options to the inliner pass to specify the max number of inlining iteration within an SCC (with a test case).

### CODE REVIEW

File: mlir-sycl/lib/Transforms/Inliner.cpp [hunk 1]
```diff
@@ -150,7 +154,7 @@ class CGUseList {
 
   /// Returns true if the given callgraph node has a single use and can be
   /// discarded.
-  bool hasOneUseAndDiscardable(CallGraphNode *CGN) const;
+  bool hasOneUse(CallGraphNode *CGN) const;
```
Comment (etiotto) [inline mlir-sycl/lib/Transforms/Inliner.cpp:157]: undo

File: mlir-sycl/lib/Transforms/Inliner.cpp [hunk 2]
```diff
@@ -407,7 +401,7 @@ bool CGUseList::isDead(CallGraphNode *CGN) const {
   return SymbolIt != DiscardableSymNodeUses.end() && SymbolIt->second == 0;
 }
 
-bool CGUseList::hasOneUseAndDiscardable(CallGraphNode *CGN) const {
+bool CGUseList::hasOneUse(CallGraphNode *CGN) const {
```
Comment (etiotto) [inline mlir-sycl/lib/Transforms/Inliner.cpp:404]: undo name change

File: mlir-sycl/lib/Transforms/Inliner.cpp [hunk 3]
```diff
@@ -509,9 +554,9 @@ LogicalResult InlinerBase::runTransformOnSCCs(
...
 }
 
-LogicalResult InlinerBase::inlineSCC(InlinerBase &Inliner, CGUseList &UseList,
-                                     CallGraphSCC &SCC,
-                                     Pass::Statistic &NumInlinedCalls) {
+LogicalResult Inliner::inlineSCC(Inliner &Inliner, CGUseList &UseList,
+                                 CallGraphSCC &SCC,
+                                 Pass::Statistic &NumInlinedCalls) {
   unsigned IterationCount = 0;
   unsigned const constexpr MaxIterationCount = 5;
```
Comment (whitneywhtsang) [inline mlir-sycl/lib/Transforms/Inliner.cpp:561]: Use the option.

**Review by whitneywhtsang — CHANGES_REQUESTED** [review]

File: mlir-sycl/test/Transforms/inliner.mlir [hunk 4]
```diff
@@ -142,3 +142,61 @@ gpu.func @gpu_func_callee() -> i32 attributes {passthrough = ["alwaysinline"]} {
 }
 
 }
+
+// -----
+
+// COM: Ensure functions in a SCC are fully inlined (requires multiple inlining iterations).
```
Comment (whitneywhtsang) [inline mlir-sycl/test/Transforms/inliner.mlir:148]: This test case doesn't always rely on multiple inlining iterations. It depends on which node in the SCC being visit first.
If you move `@callee` and `@always_inline_callee` before the other functions, then one iteration is enough.

Comment (whitneywhtsang) [reply mlir-sycl/test/Transforms/inliner.mlir:148]: Assuming we want to rely on the SCC traversal order, you can simplify this test case by removing `@inline_hint_callee1` and `@always_inline_callee`, would still need 2 iterations.

**Review by whitneywhtsang — APPROVED** [review]

File: mlir-sycl/lib/Transforms/Inliner.cpp [hunk 5]
```diff
@@ -490,10 +484,61 @@ void CGUseList::decrementDiscardableUses(CGUser &Uses) {
...
+    if (PassThroughAttr)
+      ShouldInline |= llvm::any_of(
+          PassThroughAttr->getValue().cast<ArrayAttr>(), [](Attribute Attr) {
+            return Attr.isa<StringAttr>() &&
+                   Attr.cast<StringAttr>() == "inlinehint";
+          });
+
+    // Inline a function if inlining makes it dead.
+    if (!ShouldInline)
+      ShouldInline |= Uses.hasOneUseAndDiscardable(ResolvedCall.TgtNode);
```
Comment (victor-eds) [inline mlir-sycl/lib/Transforms/Inliner.cpp:515]:
```suggestion
    ShouldInline |= Uses.hasOneUseAndDiscardable(ResolvedCall.TgtNode);
```

`|=` already shortcircuits.

File: mlir-sycl/lib/Transforms/Inliner.cpp [hunk 6]
```diff
@@ -490,10 +484,61 @@ void CGUseList::decrementDiscardableUses(CGUser &Uses) {
...
+
+  bool ShouldInline = false;
+  switch (InlineMode) {
+  case sycl::InlineMode::Aggressive:
+    llvm_unreachable("TODO");
+  case sycl::InlineMode::Simple:
+    // Inline a function if it has an attribute suggesting that inlining is
+    // desirable.
+    if (PassThroughAttr)
+      ShouldInline |= llvm::any_of(
```
Comment (victor-eds) [inline mlir-sycl/lib/Transforms/Inliner.cpp:507]:
```suggestion
      ShouldInline = llvm::any_of(
```

Will always be false.

**Review by victor-eds — CHANGES_REQUESTED** [review]

File: mlir-sycl/lib/Transforms/Inliner.cpp (hunk 1 above)
Comment (etiotto) [reply mlir-sycl/lib/Transforms/Inliner.cpp]: done

File: mlir-sycl/lib/Transforms/Inliner.cpp (hunk 2 above)
Comment (etiotto) [reply mlir-sycl/lib/Transforms/Inliner.cpp]: done

File: mlir-sycl/test/Transforms/inliner.mlir (hunk 4 above)
Comment (etiotto) [reply mlir-sycl/test/Transforms/inliner.mlir:148]: yes reduced the test

File: mlir-sycl/lib/Transforms/Inliner.cpp [hunk 7]
```diff
@@ -204,11 +208,29 @@ operator<<(llvm::raw_ostream &OS, const CGUseList &UseList) {
...
+class InlineHeuristic {
+public:
+  InlineHeuristic(sycl::InlineMode InlineMode) : InlineMode(InlineMode) {}
+
+  /// Returns true if the given call should be inlined and false otherwise.
+  bool shouldInline(ResolvedCall &ResolvedCall, const CGUseList &Uses) const;
+
+private:
+  /// Returns true if the target is an ancestor of the call and false otherwise.
+  bool isRecursiveCall(const ResolvedCall &ResolvedCall) const;
```
[the remark this answers is not in the document]
Comment (etiotto) [reply mlir-sycl/lib/Transforms/Inliner.cpp:221]: done

File: mlir-sycl/lib/Transforms/Inliner.cpp [hunk 8]
```diff
@@ -509,21 +553,20 @@ LogicalResult InlinerBase::runTransformOnSCCs(
...
   return success();
 }
 
-bool InlinerBase::inlineCallsInSCC(InlinerBase &Inliner, CGUseList &UseList,
-                                   CallGraphSCC &SCC,
-                                   Pass::Statistic &NumInlinedCalls) {
+bool Inliner::inlineCallsInSCC(Inliner &Inliner, CGUseList &UseList,
+                               CallGraphSCC &SCC,
+                               Pass::Statistic &NumInlinedCalls) {
   llvm::SmallSetVector<CallGraphNode *, 1> DeadNodes;
```
Comment (victor-eds) [inline mlir-sycl/lib/Transforms/Inliner.cpp:570]: Can we use `SmallPtrSet` here as well?

**Review by victor-eds — APPROVED** [review]
LGTM after comment.

File: mlir-sycl/lib/Transforms/Inliner.cpp (hunk 8 above)
Comment (etiotto) [reply mlir-sycl/lib/Transforms/Inliner.cpp:570]: done

**Pull request merged**